vllm/engine/llm_engine.py

```diff
@@ -792,18 +783,39 @@ def _get_default_enc_dec_decoder_prompt(self, ) -> List[int]:
 
         bos_token_id = self._get_bos_token_id()
         assert bos_token_id is not None
-        prompt_token_ids: List[int] = [bos_token_id]
-        return prompt_token_ids
+        return [bos_token_id]
+
+    def _build_enc_dec_llm_inputs(
+        self,
+        encoder_comps: PromptComponents,
+        decoder_comps: DecoderPromptComponents,
+    ) -> EncoderDecoderLLMInputs:
+        encoder_prompt, encoder_prompt_ids, encoder_mm_data = encoder_comps
+        decoder_prompt, decoder_prompt_ids, decoder_mm_data = decoder_comps
+
+        if encoder_mm_data is not None or decoder_mm_data is not None:
+            raise ValueError("Multi-modal data is not supported for "
```

Comment:
Technically the existing multi-modal models use a vision encoder, so I wanted to make it clear that "encoder" here refers to language encoder.

Reply:
The point I am about to make it not critical, don't block merging on addressing this - but just for context, the limitation is really that encoder/decoder models *with cross-attention* do not currently support multi-modal.

So for example, a model like Llava has an encoder but not cross-attention.

However, a vision model setup like the one shown in the diagram (taken from a blog post) requires cross attention between ViT and RoBERTa in order to decode a description of the image; even though this is a vision model, it would not currently be supported by vLLM because it requires both multi-modal support *and* encoder/decoder cross-attention, which is a currently-unsupported combination (for example, EncoderDecoderModelRunner has an assert which checks that multimodality is not enabled.)

![image](https://github.com/user-attachments/assets/dc22c99c-78bc-431b-9645-72e6e2c5b7d7)

Reply:
Of course this inability to have multimodal+encoder/decoder will need to be addressed by the time we add Whisper, which is a near-term goal

Reply:
Thanks for providing more context, yeah we will need to address this soon.

Reply:
I'll update the error to say "multi-modal" encoder-decoder models are not supported yet.

Reply:
Just saw this comment - yes like @afeldman-nm said, this is something we should start thinking about since Llama 3.1 multi-modal has exactly the same inference pattern like what Andrew described here.

Reply:
@ywang96 FYI I just released an RFC giving an overview of next-steps for vLLM's encoder/decoder support, and referenced this discussion in the section on multimodality support: https://github.com/vllm-project/vllm/issues/7366#user-content-considerations-for-designing-multimodal-encoderdecoder-prompt-formats